### ISSUE TITLE
Optimization for certain API Interfaces

### ISSUE THREAD

**Issue by lxcmyf** [tronprotocol/java-tron#5398]
# Rationale

Currently, java-tron's HTTP interfaces support GET and POST requests. However, there are some inconsistencies in parameter validation and response between GET and POST calls for the same interface. Here are some examples:

### case1: incorrect log level for invalid parameter

`/wallet/getapprovedlist
`
When passing an invalid parameter like modifying `type_url` to `type_url1`, the server logs an ERROR:

    18:02:57.889 ERROR [qtp464658113-329] [capsule](TransactionCapsule.java:350) Type of the Any message does not match the given class.

### case2: inconsistent naming convention for parameters

`/wallet/getavailableunfreezecount
`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/e8a4fc89-71e4-4721-bfb6-8871af81930a)

The GET and POST requests have inconsistent naming for the parameter `ownerAddress` (GET) vs `owner_address` (POST).

### case3: GET parameter handling contradicts business logic 

`/wallet/getcanwithdrawunfreezeamount
`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/cfd32929-ed20-4628-92ba-8f813d817bf1)

For the `timestamp` parameter, the GET request throws `NumberFormatException` if missing while POST handles it as 0 by default, which aligns with business logic.

`/wallet/getcandelegatedmaxsize
`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/b2b05ac6-724f-46e3-bc44-eaac488aeed5)

For the `type` parameter, GET request throws `NumberFormatException` if missing while POST handles it as 0 by default, which aligns with business logic.

`/wallet/getblockbynum
`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/a94f34fa-909a-44f3-9009-d1eb1e3c6467)
For the `num` parameter, POST resolves it to 0 without passing it, and GET throws an error. To make it compatible, GET should be consistent with POST.

### case4：Lack of compatibility handling for parameter

`/wallet/getaccount
`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/0309301a-e255-4f97-a633-9955077c460c)

The `visible` parameter for GET only works for base64 encoded addresses, lacking compatibility handling.

### case5: POST directly invokes GET

`/wallet/getnowblock`
`/wallet/listnodes`
`/wallet/getassetissuelist`
`/wallet/listproposals`
![image](https://github.com/tronprotocol/java-tron/assets/29877167/b009f599-6fc9-40db-ac89-f396407f07b6)

For these APIs, POST directly calls GET, so the `{visible\:true} `input is not handled properly.

# Implementation

To summarize, 5 cases are listed where GET and POST have inconsistencies. The solutions are as follows:

case1：change log level for invalid parameter validation

case2：standardize naming conventions

case3：align GET handling with POST logic

case4：add compatibility handling for GET parameters

case5：modify POST response based on input

**Comment by lxcmyf** [comment]
@yuekun0707 
The naming of the get request method parameters for this interface has not previously formed an example document for external users, so users rarely use unofficial incorrect parameters to call the interface. This improvement aims to strive for consistency between the get method and the post method behavior.

**Comment by xq-lu** [comment]
I can see there are both doGet() and doPost() methods in this HTTP API services. from my understanding, there are many common logic that can be used both in doGet() and doPost() methods.
can you separate this common logic and use them in a universe type, which may solve the inconsistencies between GET and POST methods?

**Comment by lxcmyf** [comment]
@xq-lu 
You have put forward a very good suggestion. I agree that this is a good idea worth considering.

**Comment by asi90sp** [comment]
> # Rationale
> Currently, java-tron's HTTP interfaces support GET and POST requests. However, there are some inconsistencies in parameter validation and response between GET and POST calls for the same interface. Here are some examples:
> 
> ### case1: incorrect log level for invalid parameter
> `/wallet/getapprovedlist ` When passing an invalid parameter like modifying `type_url` to `type_url1`, the server logs an ERROR:
> 
> ```
> 18:02:57.889 ERROR [qtp464658113-329] [capsule](TransactionCapsule.java:350) Type of the Any message does not match the given class.
> ```
> 
> ### case2: inconsistent naming convention for parameters
> `/wallet/getavailableunfreezecount ` ![image](https://user-images.githubusercontent.com/29877167/258058074-e8a4fc89-71e4-4721-bfb6-8871af81930a.png)
> 
> The GET and POST requests have inconsistent naming for the parameter `ownerAddress` (GET) vs `owner_address` (POST).
> 
> ### case3: GET parameter handling contradicts business logic 
> `/wallet/getcanwithdrawunfreezeamount ` ![image](https://user-images.githubusercontent.com/29877167/258058174-cfd32929-ed20-4628-92ba-8f813d817bf1.png)
> 
> For the `timestamp` parameter, the GET request throws `NumberFormatException` if missing while POST handles it as 0 by default, which aligns with business logic.
> 
> `/wallet/getcandelegatedmaxsize ` ![image](https://user-images.githubusercontent.com/29877167/258058341-b2b05ac6-724f-46e3-bc44-eaac488aeed5.png)
> 
> For the `type` parameter, GET request throws `NumberFormatException` if missing while POST handles it as 0 by default, which aligns with business logic.
> 
> ### case4：Lack of compatibility handling for parameter
> `/wallet/getaccount ` ![image](https://user-images.githubusercontent.com/29877167/258058425-0309301a-e255-4f97-a633-9955077c460c.png)
> 
> The `visible` parameter for GET only works for base64 encoded addresses, lacking compatibility handling.
> 
> ### case5: POST parameter handling contradicts business logic
> `/wallet/getblockbynum ` ![image](https://user-images.githubusercontent.com/29877167/258058530-a94f34fa-909a-44f3-9009-d1eb1e3c6467.png)
> 
> For the `num` parameter, POST handles it as 0 if missing while GET throws an error, which aligns with business logic.
> 
> ### case6: POST directly invokes GET
> `/wallet/getnowblock` `/wallet/listnodes` `/wallet/getassetissuelist` `/wallet/listproposals` ![image](https://user-images.githubusercontent.com/29877167/258058735-b009f599-6fc9-40db-ac89-f396407f07b6.png)
> 
> For these APIs, POST directly calls GET, so the `{visible\:true} `input is not handled properly.
> 
> ### case7: GET default handling misaligns with business logic
> `/wallet/getBrokerage` `/wallet/getReward` ![image](https://user-images.githubusercontent.com/29877167/258058862-65be92f9-354e-47fb-a9dc-38e37897f13b.png)
> 
> For `address` parameter, the default 0 return value contradicts the logic and should throw an error.
> 
> # Implementation
> To summarize, 7 cases are listed where GET and POST have inconsistencies. The solutions are as follows:
> 
> case1：change log level for invalid parameter validation
> 
> case2：standardize naming conventions
> 
> case3：align GET handling with POST logic
> 
> case4：add compatibility handling for GET parameters
> 
> case5：validate POST parameters to match GET logic
> 
> case6：modify POST response based on input
> 
> case7：throw error for invalid GET parameters

Ok

**Comment by halibobo1205** [comment]
Keep in mind to keep compatibility.

**Comment by halibobo1205** [comment]
Will application/x-www-form-urlencoded be supported?

**Comment by lxcmyf** [comment]
@halibobo1205 
> Keep in mind to keep compatibility.

Of course.

> Will `application/x-www-form-urlencoded` be supported?

Yes, it will be compatible with this content-type.
